File: shared_resources/src/main/java/com/thewizrd/shared_resources/weatherdata/images/ImageDataHelperImpl.java
package com.thewizrd.shared_resources.weatherdata.images;

import android.net.Uri;

import androidx.annotation.WorkerThread;

import com.thewizrd.shared_resources.weatherdata.Weather;
import com.thewizrd.shared_resources.weatherdata.images.model.ImageData;

public abstract class ImageDataHelperImpl {
    @WorkerThread
    public abstract ImageData getCachedImageData(String backgroundCode);

    @WorkerThread
    public ImageData getRemoteImageData(String backgroundCode) {
        ImageData imageData = ImageDatabase.getRandomImageForCondition(backgroundCode);

        if (imageData != null && imageData.isValid()) {
            ImageData cachedImage = cacheImage(imageData);
            return cachedImage;
        }

        return null;
    }

    @WorkerThread
    public ImageData cacheImage(ImageData imageData) {
        Uri imageUri = Uri.parse(imageData.getImageURL());
        if ("gs".equals(imageUri.getScheme()) || "https".equals(imageUri.getScheme()) || "http".equals(imageUri.getScheme())) {
            // Download image to storage
            // and image metadata to settings
            ImageData cachedImage = storeImage(imageUri, imageData);
            return cachedImage;
        }

        // Invalid image uri
        return null;
    }

    @WorkerThread
    protected abstract ImageData storeImage(Uri imageUri, ImageData imageData);

    @WorkerThread
    public abstract void clearCachedImageData();

    @WorkerThread
    public abstract ImageData getDefaultImageData(String backgroundCode, Weather weather);

    @WorkerThread
    public abstract boolean isEmpty();
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/weatherdata/images/ImageDataHelperImpl.kt
package com.thewizrd.shared_resources.weatherdata.images

import android.net.Uri
import androidx.annotation.WorkerThread
import com.thewizrd.shared_resources.weatherdata.Weather
import com.thewizrd.shared_resources.weatherdata.images.model.ImageData

abstract class ImageDataHelperImpl {
    @WorkerThread
    abstract suspend fun getCachedImageData(backgroundCode: String?): ImageData?

    @WorkerThread
    suspend fun getRemoteImageData(backgroundCode: String?): ImageData? {
        val imageData = ImageDatabase.getRandomImageForCondition(backgroundCode)

        return if (imageData?.isValid == true) {
            cacheImage(imageData)
        } else {
            null
        }
    }

    @WorkerThread
    suspend fun cacheImage(imageData: ImageData): ImageData? {
        val imageUri = Uri.parse(imageData.imageURL)
        return if ("gs" == imageUri.scheme || "https" == imageUri.scheme || "http" == imageUri.scheme) {
            // Download image to storage
            // and image metadata to settings
            storeImage(imageUri, imageData)
        } else {
            // Invalid image uri
            null
        }
    }

    @WorkerThread
    protected abstract suspend fun storeImage(imageUri: Uri, imageData: ImageData): ImageData?

    @WorkerThread
    abstract suspend fun clearCachedImageData()

    @WorkerThread
    abstract suspend fun getDefaultImageData(backgroundCode: String?, weather: Weather): ImageData?

    @get:WorkerThread
    abstract val isEmpty: Boolean
}